# Day 4: Giant Squid

## Part one

You're already almost 1.5km (almost a mile) below the surface of the ocean, already so deep that you can't see any sunlight. What you can see, however, is a giant squid that has attached itself to the outside of your submarine.

Maybe it wants to play bingo?

Bingo is played on a set of boards each consisting of a 5x5 grid of numbers. Numbers are chosen at random, and the chosen number is marked on all boards on which it appears. (Numbers may not appear on all boards.) If all numbers in any row or any column of a board are marked, that board wins. (Diagonals don't count.)

The submarine has a bingo subsystem to help passengers (currently, you and the giant squid) pass the time. It automatically generates a random order in which to draw numbers and a random set of boards (your puzzle input). For example:

`7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1`

```
22 13 17 11  0
 8  2 23  4 24
21  9 14 16  7
 6 10  3 18  5
 1 12 20 15 19

 3 15  0  2 22
 9 18 13 17  5
19  8  7 25 23
20 11 10 24  4
14 21 16 12  6

14 21 17 24  4
10 16 15  9 19
18  8 23 26 20
22 11 13  6  5
 2  0 12  3  7
```

After the first five numbers are drawn (`7`, `4`, `9`, `5`, and `11`), there are no winners, but the boards are marked as follows (shown here adjacent to each other to save space):

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

After the next six numbers are drawn (`17`, `23`, `2`, `0`, `14`, and `21`), there are still no winners:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

Finally, `24` is drawn:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

At this point, the third board **wins** because it has at least one complete row or column of marked numbers (in this case, the entire top row is marked: `14 21 17 24 4`).

The **score** of the winning board can now be calculated. Start by finding the **sum of all unmarked numbers** on that board; in this case, the sum is `188`. Then, multiply that sum by **the number that was just called** when the board won, `24`, to get the final score, `188 * 24 = 4512`.

To guarantee victory against the giant squid, figure out which board will win first. **What will your final score be if you choose that board?**

In [106]:
with open('data/input.txt', 'r') as f:
    lines = f.read().splitlines()
    
draws = lines[0].split(',')

def build_board(txt_lines):
    board = []
    
    for txt_line in txt_lines:
        line = []
        values = txt_line.split(' ')
        for value in values:
            if not value:
                continue
            item = {}
            item["value"] = int(value)
            item["drawn"] = False
            line.append(item)
        board.append(line)
    
    return board

line_number = 2
boards = []

while (line_number < len(lines)):
    board = build_board(lines[line_number:line_number+5])
    boards.append(board)
    line_number += 6
    
boards[:1]

[[[{'value': 66, 'drawn': False},
   {'value': 78, 'drawn': False},
   {'value': 7, 'drawn': False},
   {'value': 45, 'drawn': False},
   {'value': 92, 'drawn': False}],
  [{'value': 39, 'drawn': False},
   {'value': 38, 'drawn': False},
   {'value': 62, 'drawn': False},
   {'value': 81, 'drawn': False},
   {'value': 77, 'drawn': False}],
  [{'value': 9, 'drawn': False},
   {'value': 73, 'drawn': False},
   {'value': 25, 'drawn': False},
   {'value': 97, 'drawn': False},
   {'value': 99, 'drawn': False}],
  [{'value': 87, 'drawn': False},
   {'value': 80, 'drawn': False},
   {'value': 19, 'drawn': False},
   {'value': 1, 'drawn': False},
   {'value': 71, 'drawn': False}],
  [{'value': 85, 'drawn': False},
   {'value': 35, 'drawn': False},
   {'value': 52, 'drawn': False},
   {'value': 26, 'drawn': False},
   {'value': 68, 'drawn': False}]]]

In [107]:
def bingo_draw(selected_number, boards):
    game_status = {'board': 0, 'win': False}
    
    for board_number in range(0, len(boards)):
        board = boards[board_number]
        for line_number in range(0, len(board)):
            line = board[line_number]
            for col_number in range(0, len(line)):
                box = line[col_number]
                if selected_number == box['value']:
                    box['drawn'] = True
                    
                    if bingo_win(board, line_number, col_number):
                        game_status['board'] = board_number
                        game_status['win'] = True
                        break
                        
    return game_status
                    
def bingo_win(board, line_number, col_number):
    win = True
    
    for x in range(0, 5):
        if board[x][col_number]['drawn'] == False:
            win = False
            break
    
    if win == False:
        win = True
        for y in range(0, 5):
            if board[line_number][y]['drawn'] == False:
                win = False
                break
    
    return win

In [108]:
for selected_number in draws:
    selected_number = int(selected_number)
    game_status = bingo_draw(selected_number, boards)
    
    if game_status['win'] == True:
        break
        
print("board_number: %s - selected_number: %s" % (game_status['board'], selected_number, ))
boards[game_status['board']]

board_number: 14 - selected_number: 54


[[{'value': 30, 'drawn': True},
  {'value': 93, 'drawn': False},
  {'value': 48, 'drawn': False},
  {'value': 17, 'drawn': False},
  {'value': 33, 'drawn': False}],
 [{'value': 67, 'drawn': True},
  {'value': 7, 'drawn': False},
  {'value': 5, 'drawn': False},
  {'value': 0, 'drawn': False},
  {'value': 69, 'drawn': False}],
 [{'value': 54, 'drawn': True},
  {'value': 76, 'drawn': True},
  {'value': 52, 'drawn': False},
  {'value': 1, 'drawn': False},
  {'value': 87, 'drawn': True}],
 [{'value': 99, 'drawn': True},
  {'value': 73, 'drawn': False},
  {'value': 50, 'drawn': True},
  {'value': 25, 'drawn': False},
  {'value': 16, 'drawn': False}],
 [{'value': 13, 'drawn': True},
  {'value': 20, 'drawn': False},
  {'value': 41, 'drawn': False},
  {'value': 77, 'drawn': False},
  {'value': 62, 'drawn': False}]]

In [109]:
def sum_unmarked(board):
    total = 0
    for line in board:
        for box in line:
            if box['drawn'] == False:
                total += box['value']
                
    return total

In [110]:
sum_unmarked(boards[game_status['board']]) * selected_number

34506

Your puzzle answer was `34506`.

## Part two

On the other hand, it might be wise to try a different strategy: let the giant squid win.

You aren't sure how many bingo boards a giant squid could play at once, so rather than waste time counting its arms, the safe thing to do is to **figure out which board will win last** and choose that one. That way, no matter which boards it picks, it will win for sure.

In the above example, the second board is the last to win, which happens after 13 is eventually called and its middle column is completely marked. If you were to keep playing until this point, the second board would have a sum of unmarked numbers equal to 148 for a final score of `148 * 13 = 1924`.

Figure out which board will win last. **Once it wins, what would its final score be?**

In [112]:
with open('data/input.txt', 'r') as f:
    lines = f.read().splitlines()
    
draws = lines[0].split(',')

def build_board(txt_lines):
    board = []
    
    for txt_line in txt_lines:
        line = []
        values = txt_line.split(' ')
        for value in values:
            if not value:
                continue
            item = {}
            item["value"] = int(value)
            item["drawn"] = False
            line.append(item)
        board.append(line)
    
    return board

line_number = 2
boards = []

while (line_number < len(lines)):
    board = build_board(lines[line_number:line_number+5])
    boards.append(board)
    line_number += 6
    
boards[:1]

[[[{'value': 66, 'drawn': False},
   {'value': 78, 'drawn': False},
   {'value': 7, 'drawn': False},
   {'value': 45, 'drawn': False},
   {'value': 92, 'drawn': False}],
  [{'value': 39, 'drawn': False},
   {'value': 38, 'drawn': False},
   {'value': 62, 'drawn': False},
   {'value': 81, 'drawn': False},
   {'value': 77, 'drawn': False}],
  [{'value': 9, 'drawn': False},
   {'value': 73, 'drawn': False},
   {'value': 25, 'drawn': False},
   {'value': 97, 'drawn': False},
   {'value': 99, 'drawn': False}],
  [{'value': 87, 'drawn': False},
   {'value': 80, 'drawn': False},
   {'value': 19, 'drawn': False},
   {'value': 1, 'drawn': False},
   {'value': 71, 'drawn': False}],
  [{'value': 85, 'drawn': False},
   {'value': 35, 'drawn': False},
   {'value': 52, 'drawn': False},
   {'value': 26, 'drawn': False},
   {'value': 68, 'drawn': False}]]]

In [113]:
def bingo_draw_two(selected_number, boards, ended_boards):
    game_status = {'board': 0, 'ended': False}
    
    for board_number in range(0, len(boards)):
        board = boards[board_number]
        if board_number in ended_boards:
            continue
        for line_number in range(0, len(board)):
            line = board[line_number]
            for col_number in range(0, len(line)):
                box = line[col_number]
                if selected_number == box['value']:
                    box['drawn'] = True
                    
                    if bingo_win(board, line_number, col_number):
                        game_status['board'] = board_number
                        
                        ended_boards.append(board_number)
                        
                        if len(ended_boards) == len(boards):
                            game_status['ended'] = True
                            break
    
    return game_status

In [114]:
ended_boards = []

for selected_number in draws:
    selected_number = int(selected_number)
    game_status = bingo_draw_two(selected_number, boards, ended_boards)
    
    if game_status['ended'] == True:
        break
        
print("board_number: %s - selected_number: %s" % (game_status['board'], selected_number, ))
boards[game_status['board']]

board_number: 58 - selected_number: 42


[[{'value': 93, 'drawn': False},
  {'value': 7, 'drawn': True},
  {'value': 60, 'drawn': True},
  {'value': 75, 'drawn': True},
  {'value': 12, 'drawn': True}],
 [{'value': 49, 'drawn': True},
  {'value': 64, 'drawn': True},
  {'value': 20, 'drawn': False},
  {'value': 46, 'drawn': True},
  {'value': 10, 'drawn': True}],
 [{'value': 3, 'drawn': True},
  {'value': 23, 'drawn': True},
  {'value': 76, 'drawn': True},
  {'value': 42, 'drawn': True},
  {'value': 47, 'drawn': True}],
 [{'value': 9, 'drawn': True},
  {'value': 22, 'drawn': False},
  {'value': 6, 'drawn': True},
  {'value': 34, 'drawn': True},
  {'value': 87, 'drawn': True}],
 [{'value': 41, 'drawn': True},
  {'value': 37, 'drawn': True},
  {'value': 66, 'drawn': True},
  {'value': 45, 'drawn': True},
  {'value': 48, 'drawn': False}]]

In [115]:
sum_unmarked(boards[game_status['board']]) * selected_number

7686

Your puzzle answer was 7686.